python/dataproc_templates/bigquery/bigquery_to_gcs.py

```diff
@@ -0,0 +1,119 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Optional, Any
+import argparse
+import pprint
+
+from pyspark.sql import SparkSession, DataFrameWriter
+
+from dataproc_templates import BaseTemplate
+import dataproc_templates.util.template_constants as constants
+
+__all__ = ['BigQueryToGCSTemplate']
+
+
+class BigQueryToGCSTemplate(BaseTemplate):
+    """
+    Dataproc template implementing exports from BigQuery to GCS
+    """
+
+    @staticmethod
+    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
+        parser: argparse.ArgumentParser = argparse.ArgumentParser()
+
+        parser.add_argument(
+            f'--{constants.BQ_GCS_INPUT_TABLE}',
+            dest=constants.BQ_GCS_INPUT_TABLE,
+            required=True,
+            help='BigQuery Input table name'
+        )
+        parser.add_argument(
+            f'--{constants.BQ_GCS_OUTPUT_FORMAT}',
+            dest=constants.BQ_GCS_OUTPUT_FORMAT,
+            required=True,
+            help='Output file format (one of: avro,parquet,csv,json)',
+            choices=[
+                constants.FORMAT_AVRO,
+                constants.FORMAT_PRQT,
+                constants.FORMAT_CSV,
+                constants.FORMAT_JSON
+            ]
+        )
+        parser.add_argument(
+            f'--{constants.BQ_GCS_OUTPUT_LOCATION}',
+            dest=constants.BQ_GCS_OUTPUT_LOCATION,
+            required=True,
+            help='GCS location for output files'
+        )
+        parser.add_argument(
+            f'--{constants.BQ_GCS_OUTPUT_MODE}',
+            dest=constants.BQ_GCS_OUTPUT_MODE,
+            required=False,
+            default=constants.OUTPUT_MODE_APPEND,
+            help=(
+                'Output write mode '
+                '(one of: append,overwrite,ignore,errorifexists) '
+                '(Defaults to append)'
+            ),
+            choices=[
+                constants.OUTPUT_MODE_OVERWRITE,
+                constants.OUTPUT_MODE_APPEND,
+                constants.OUTPUT_MODE_IGNORE,
+                constants.OUTPUT_MODE_ERRORIFEXISTS
+            ]
+        )
+
+        known_args: argparse.Namespace
+        known_args, _ = parser.parse_known_args(args)
+
+        return vars(known_args)
+
+    def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:
+
+        log4jLogger = spark.sparkContext._jvm.org.apache.log4j  # pylint: disable=protected-access,invalid-name
```

Comment:
is pylint comment supposed to be here ?

Reply:
It is. By default, pylint flags accessing protected fields in an object (anything starting with an underscore) as an error and warns you about it. However; getting the Spark logger from Python requires accessing `SparkContext._jvm`, which triggers the warning.

With that comment, we acknowledge we're accessing a protected field and disable the `protected-access` check only for that line. That way; accessing `_jvm` does not trigger a linter warning, but we get to keep that check for the rest of the project.

We're also disabling `invalid-name`, but that avoid that by just renaming the variable to use snake case. Will do that later!